keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resolvers/NetworkResolver.kt

```diff
@@ -0,0 +1,87 @@
+package com.netflix.spinnaker.keel.ec2.resolvers
+
+import com.netflix.spinnaker.keel.api.ApiVersion
+import com.netflix.spinnaker.keel.api.Locatable
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.SimpleLocations
+import com.netflix.spinnaker.keel.api.SubnetAwareLocations
+import com.netflix.spinnaker.keel.api.ec2.ApplicationLoadBalancerSpec
+import com.netflix.spinnaker.keel.api.ec2.ClassicLoadBalancerSpec
+import com.netflix.spinnaker.keel.api.ec2.ClusterSpec
+import com.netflix.spinnaker.keel.ec2.SPINNAKER_EC2_API_V1
+import com.netflix.spinnaker.keel.plugin.Resolver
+import org.springframework.stereotype.Component
+
+abstract class NetworkResolver<T : Locatable<*>> : Resolver<T> {
+
+  protected fun SubnetAwareLocations.withResolvedNetwork(): SubnetAwareLocations {
```

Comment:
Just curious: is it a common pattern to have a class whose sole purpose is to declare extensions for a different class like you did here? It wasn't initially clear to me that `NetworkResolver` types were being used anywhere in the code until I went back and noticed the `withResolvedNetwork` methods and the overrides to `invoke` again. 
Also, it looks like all 3 sub-types of `NetworkResolver` have the exact same override for `invoke` and the only difference is they have a different `supportedKind` but I couldn't spot where that was used either. I guess in general I was just wondering if we could have just added the code to resolve locations to the base `ResourceSpec` class (although I do realize not all resources are "locatable", but maybe in the `Locatable` interface then?) to simplify things.

Reply:
It's because `Resolver<T>` extends the functional type `(Resource<T>) -> Resource<T>` which specifies the `invoke` method. That's an operator function. If an object has `invoke` you can use it like a function: e.g. `myThing("foo")`.